yarn-project/telemetry-client/src/prom_otel_adapter.ts

```diff
@@ -197,84 +198,215 @@ export class OtelGauge<Labels extends LabelsGeneric = NoLabels> implements IGaug
-  observe(_labels: Labels, _value: number): void;
-  observe(_valueOrLabels: number | Labels, _value?: number): void {}
-
-  startTimer(_labels?: Labels): (_labels?: Labels) => number {
-    return () => 0;
+    meter: Meter,
+    name: string,
+    help: string,
+    private labelNames: Array<keyof Labels> = [],
+  ) {
+    // Create three separate gauges for avg, min, and max
+    this.gauges = {
+      avg: meter.createObservableGauge(`${name}_avg` as MetricsType, {
+        description: `${help} (average)`,
+      }),
+      min: meter.createObservableGauge(`${name}_min` as MetricsType, {
+        description: `${help} (minimum)`,
+      }),
+      max: meter.createObservableGauge(`${name}_max` as MetricsType, {
+        description: `${help} (maximum)`,
+      }),
+    };
+
+    // Register callbacks for each gauge
+    this.gauges.avg.addCallback(this.observeAvg.bind(this));
+    this.gauges.min.addCallback(this.observeMin.bind(this));
+    this.gauges.max.addCallback(this.observeMax.bind(this));
   }
 
+  /**
+   * Sets the values for calculating avg, min, max
+   * @param values - Array of values
+   */
+  set(values: number[]): void;
+  /**
+   * Sets the values for calculating avg, min, max with labels
+   * @param labels - Labels object
+   * @param values - Array of values
+   */
+  set(labels: Labels, values: number[]): void;
+  set(labelsOrValues: number[] | Labels, values?: number[]): void {
+    if (Array.isArray(labelsOrValues)) {
+      this.currentValues = labelsOrValues;
+      return;
+    } else {
+      validateLabels(labelsOrValues, this.labelNames, 'AvgMinMax');
+      const labelKey = JSON.stringify(labelsOrValues);
+      this.labeledValues.set(labelKey, values || []);
+    }
+  }
+
+  /**
+   * Resets all stored values
+   */
   reset(): void {
-    // OpenTelemetry histograms cannot be reset, but we implement the interface
-    this.logger.silent('OpenTelemetry histograms cannot be reset');
+    this.currentValues = [];
+    this.labeledValues.clear();
+  }
+
+  /**
+   * General function to observe an aggregation
+   * @param result - Observer result
+   * @param aggregateFn - Function that calculates the aggregation
+   */
+  private observeAggregation(result: any, aggregateFn: (arr: number[]) => number): void {
+    // Observe unlabeled values
+    if (this.currentValues.length > 0) {
+      result.observe(aggregateFn(this.currentValues));
+    }
+
+    // Observe labeled values
+    for (const [labelStr, values] of this.labeledValues.entries()) {
+      if (values.length > 0) {
+        const labels = parseLabelsSafely(labelStr, this.logger);
+        if (labels) {
+          result.observe(aggregateFn(values), labels);
+        }
+      }
+    }
+  }
+
+  private observeAvg(result: any): void {
+    this.observeAggregation(result, arr => arr.reduce((sum, val) => sum + val, 0) / arr.length);
+  }
+
+  private observeMin(result: any): void {
+    this.observeAggregation(result, arr => Math.min(...arr));
```

Comment:
```suggestion
    this.observeAggregation(result, arr => Math.min.apply(null, arr));
```

I was thinking that arr here can potentially be of unbounded size, so the array copy here could also be unbounded, it could be preferential to pass by reference here.


Alternatively, rather than recomputing the Avg, Max and Min on each call, for all we can store the current state for each, then do a small recomputation on each call. 

Min: is the new value lower, if so update
Max: is the new value larger, if so update
Avg: ((runningAvg * |obserations|) + new value) / |observations + 1| ( or store a running total + count )